parson.c

```diff
@@ -486,12 +488,39 @@ static JSON_Status skip_quotes(const char **string) {
     return JSONSuccess;
 }
 
+static int scan_hex(const char c) {
+    if (c >= '0' && c <= '9') {
+        return c - '0';
+    } else if (c >= 'A' && c <= 'F') {
+        return c - 'A' + 10;
+    } else if (c >= 'a' && c <= 'f') {
+        return c - 'a' + 10;
+    } else {
+        return EOF;
+    }
+}
+
+static int scan_utf_16(const char *str, unsigned int *cp) {
+    unsigned int cp1 = scan_hex(*str++) << 12;
+    unsigned int cp2 = scan_hex(*str++) << 8;
+    unsigned int cp3 = scan_hex(*str++) << 4;
+    unsigned int cp4 = scan_hex(*str);
+
+    if (cp1 == EOF || cp2 == EOF || cp3 == EOF || cp4 == EOF) {
```

Comment:
It should be compared to EOF **before** being shifted (thanks to @mbelicki for finding this bug). Also, unsigned int shouldn't be compared with an int (EOF).